### ISSUE TITLE
TypeScript does not infer excess property check for union of optional keys in object literal when using computed property names

### ISSUE THREAD

**Issue by patrick-bricout-spendesk** [microsoft/TypeScript#60274]
🕗 Version & Regression Information

This is the behavior in every version I tried in TS playground.

⏯ Playground Link

https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABMAFDMAHEUBciDeiA1gKYCeAjAPx4DOUATugOYC+AlAQFCtdcD0-RACESACwCGAN3ggGiCGJIQieKEoYlEMWoglISADwzKoJACaISDBnHkB3GOsTqtwOABsPcR2GYKJLy5gfHxSMgAmPAAiKUCQEmjWXmDwaHgkYAiUcJjwimjOfC5EUuQUfABtcIBdGLiPBKT2Hj5QSFgEZABmHPI88gjC7jLEQUQAQV0jE2gLFyVkT29ffwhAj2QJGA9dR2d9Kxs7HBKy8YB5ACMAK1NEDydrQMQAWwkyRAQPT9pZmGAnyIYB8SAwthMDFgJFoABo9GBLAByarkGpIxDmOAwxAgqBWQw6fHoFxkEyIJGEfI0RD0Jh+RAAH0Q4HMJGA6AsAG5EKwkQA6M6lVBVWr1eKJDitVIdDLIAAsfTIA0o0SZiGi4SGRSFYyEABVFu4vD4WA9ObSxHAQB5LJoMHZ8YdrLZ5I9SNpMt1EFdsJjsbo8bqRaiyHUNQ0mlLeEA

💻 Code

```ts
function f(input: { key1?: string}) {
}

// Behaviour check: there is an expected error with the following call
f{{key2: "value"}}

function f2(key: "key1") {
    f({[key]: "value"})
}

function f3(key: "key2") {
    // As expected the following call fails with an error:
    // Object literal may only specify known properties, and '[key]' does not exist in type '{ key1?: string | undefined; }'.
    f({[key]: "value"})
}

function f4(key: "key1" | "key2") {
    // The following line should report an error like in f3 but does not
    f({[key]: "value"})
}
```

🙁 Actual behavior

TypeScript allows `f({[key]: "value"})` in `f4(key: "key1" | "key2")` without an error, even when `key` is `"key2"`, which isn't a valid property in the `input` type.

🙂 Expected behavior

TypeScript should report an error in `f4` when `key` is `"key2"`, just like it does in `f3`, since `"key2"` is not a known property of the type `{ key1?: string }`.

**Comment by MartinJohns** [comment]
Related or possibly even a duplicate: #36920

**Comment by RyanCavanaugh** [comment]
This isn't a bug. Excess property checks are there to ensure that you don't *statically* pass a misnamed property, but this property is possibly valid, therefore not excess.
